lib/dirac_coarse.cpp

```diff
@@ -190,21 +193,81 @@ namespace quda {
   {
     createY(gpu_setup, mapped);
 
-    if (gpu_setup) dirac->createCoarseOp(*Y_d,*X_d,*transfer,kappa,mass,Mu(),MuFactor());
-    else dirac->createCoarseOp(*Y_h,*X_h,*transfer,kappa,mass,Mu(),MuFactor());
+    if (!gpu_setup) {
 
-    // save the intermediate tunecache after the UV and VUV tune
-    saveTuneCache();
+      dirac->createCoarseOp(*Y_h, *X_h, *transfer, kappa, mass, Mu(), MuFactor());
+      // save the intermediate tunecache after the UV and VUV tune
+      saveTuneCache();
+      if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to build the preconditioned coarse clover\n");
+
+      createYhat(gpu_setup);
+
+      if (getVerbosity() >= QUDA_VERBOSE) printfQuda("Finished building the preconditioned coarse clover\n");
+      if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to create the preconditioned coarse op\n");
+
+      calculateYhat(*Yhat_h, *Xinv_h, *Y_h, *X_h, use_mma);
+
+    } else {
+
+      // The following fancy copies reduce the number of gauge field
+      // copies (from and to QUDA_MILC_GAUGE_ORDER) by 2: one for X
+      // and one for Y, both to QUDA_MILC_GAUGE_ORDER.
+      if (use_mma && dirac->isCoarse()) {
+
+        constexpr QudaGaugeFieldOrder gOrder = QUDA_MILC_GAUGE_ORDER;
+
+        GaugeFieldParam Y_param(*Y_d);
+        GaugeFieldParam X_param(*X_d);
+
+        Y_param.order = gOrder;
+        X_param.order = gOrder;
+
+        GaugeField *Y_order = cudaGaugeField::Create(Y_param);
+        GaugeField *X_order = cudaGaugeField::Create(X_param);
+
+        dirac->createCoarseOp(*Y_order, *X_order, *transfer, kappa, mass, Mu(), MuFactor());
+
+        // save the intermediate tunecache after the UV and VUV tune
+        saveTuneCache();
 
-    if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to build the preconditioned coarse clover\n");
+        X_d->copy(*X_order);
 
-    createYhat(gpu_setup);
+        if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to build the preconditioned coarse clover\n");
 
-    if (getVerbosity() >= QUDA_VERBOSE) printfQuda("Finished building the preconditioned coarse clover\n");
-    if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to create the preconditioned coarse op\n");
+        createYhat(gpu_setup);
 
-    if (gpu_setup) createPreconditionedCoarseOp(*Yhat_d,*Xinv_d,*Y_d,*X_d);
-    else createPreconditionedCoarseOp(*Yhat_h,*Xinv_h,*Y_h,*X_h);
+        if (getVerbosity() >= QUDA_VERBOSE) printfQuda("Finished building the preconditioned coarse clover\n");
+        if (getVerbosity() >= QUDA_VERBOSE) printfQuda("About to create the preconditioned coarse op\n");
+
+        calculateYhat(*Yhat_d, *Xinv_d, *Y_order, *X_order, use_mma);
+
+        Y_d->copy(*Y_order);
+
+        // this extra exchange shouldn't be needed, but at present the
+        // copy from Y_order to Y_d doesn't preserve the
+        // bi-directional halo (in_offset isn't set in the copy
+        // routine)
+        Y_d->exchangeGhost(QUDA_LINK_BIDIRECTIONAL);
```

Comment:
With the latest bugfixes can we get rid of this extra halo exchange? @maddyscientist

Reply:
Discussed offline, we cannot. For a future PR.